quinn-proto/src/crypto/ring.rs

```diff
@@ -231,40 +227,35 @@ fn header_key_from_secret(aead: &aead::Algorithm, secret_key: &SigningKey) -> He
     }
 }
 
-fn expanded_initial_secret(prk: &SigningKey, label: &[u8]) -> Vec<u8> {
+fn expanded_initial_secret(prk: &hkdf::Prk, label: &[u8]) -> Vec<u8> {
     let mut out = vec![0u8; digest::SHA256.output_len];
     hkdf_expand(prk, label, &mut out);
     out
 }
 
-fn hkdf_expand(key: &SigningKey, label: &[u8], out: &mut [u8]) {
-    let mut info = Vec::with_capacity(2 + 1 + 5 + out.len());
-    info.put_u16_be(out.len() as u16);
+fn hkdf_expand(key: &hkdf::Prk, label: &[u8], out: &mut [u8]) {
+    let out_len = (out.len() as u16).to_be_bytes();
     const BASE_LABEL: &[u8] = b"tls13 ";
-    info.put_u8((BASE_LABEL.len() + label.len()) as u8);
-    info.extend_from_slice(BASE_LABEL);
-    info.extend_from_slice(&label);
-    info.put_u8(0);
-    hkdf::expand(key, &info, out);
+    let label_len = (BASE_LABEL.len() + label.len()) as u8;
+    key.expand(&[&out_len, &[label_len][..], BASE_LABEL, label, &[0][..]], HKDF_SHA256).unwrap().fill(out);
 }
 
-fn initial_secret(conn_id: &ConnectionId) -> SigningKey {
-    let key = SigningKey::new(&digest::SHA256, &INITIAL_SALT);
-    hkdf::extract(&key, conn_id)
+fn initial_secret(conn_id: &ConnectionId) -> hkdf::Prk {
+    hkdf::Salt::new(HKDF_SHA256, &INITIAL_SALT).extract(conn_id)
 }
 
 const INITIAL_SALT: [u8; 20] = [
     0x7f, 0xbc, 0xdb, 0x0e, 0x7c, 0x66, 0xbb, 0xe9, 0x19, 0x3a, 0x96, 0xcd, 0x21, 0x51, 0x9e, 0xbd,
     0x7a, 0x02, 0x64, 0x4a,
 ];
 
-impl crypto::HmacKey for hmac::SigningKey {
+impl crypto::HmacKey for hmac::Key {
     const KEY_LEN: usize = 64;
-    type Signature = hmac::Signature;
+    type Signature = hmac::Tag;
 
     fn new(key: &[u8]) -> Result<Self, ConfigError> {
         if key.len() == Self::KEY_LEN {
-            Ok(hmac::SigningKey::new(&digest::SHA512_256, key))
+            Ok(hmac::Key::new(HMAC_SHA256, key))
```

Comment:
I didn't realize anybody was using HMAC-SHA-512/256. We could bring it back but I'm curious why you were originally using it.

Reply:
We favored it on the understanding that it's substantially faster than SHA-256. Use of this HMAC is not expected to be a bottleneck under normal circumstances. It might matter to a server suffering a certain type of DoS attack, but in that case we might be better off pursuing a blake2 HMAC, so I think switching to whatever *ring* prefers to offer for now is fine.